fontir/src/piecewise_linear_map.rs

```diff
@@ -0,0 +1,112 @@
+//!
+//! Useful for things like designspace : userspace mapping.
+
+use ordered_float::OrderedFloat;
+
+pub struct PiecewiseLinearMap {
+    from: Vec<OrderedFloat<f32>>, // sorted, ||'s to
+    to: Vec<OrderedFloat<f32>>,   // sorted, ||'s from
+}
+
+impl PiecewiseLinearMap {
+    /// Create a new map from a series of (from, to) values.
+    pub fn new(mut mappings: Vec<(OrderedFloat<f32>, OrderedFloat<f32>)>) -> PiecewiseLinearMap {
+        mappings.sort();
+        let (from, to): (Vec<_>, Vec<_>) = mappings.into_iter().unzip();
+        PiecewiseLinearMap { from, to }
+    }
+
+    pub fn reverse(&self) -> PiecewiseLinearMap {
+        let mappings = self
+            .to
+            .iter()
+            .copied()
+            .zip(self.from.iter().copied())
+            .collect();
+        PiecewiseLinearMap::new(mappings)
+    }
+}
+
+fn lerp(a: f32, b: f32, t: f32) -> f32 {
+    assert!((0_f32..=1_f32).contains(&t));
+    a + t * (b - a)
+}
+
+impl PiecewiseLinearMap {
+    /// Based on <https://github.com/fonttools/fonttools/blob/5a0dc4bc8dfaa0c7da146cf902395f748b3cebe5/Lib/fontTools/varLib/models.py#L502>
+    pub fn map(&self, value: OrderedFloat<f32>) -> OrderedFloat<f32> {
+        match self.from.binary_search(&value) {
+            Ok(idx) => self.to[idx], // This value is just right
+            Err(idx) => {
+                // idx is where we would insert from.
+                // Interpolate between the values left/right of it, if any.
+                let value = value.into_inner();
+
+                // This value is too small
+                if idx == 0 {
+                    // FontTools: take min k and compute v + mapping[k] - k
+                    return (value + self.to[0].into_inner() - self.from[0].into_inner()).into();
+                }
+                // This value is too big
+                if idx == self.from.len() {
+                    // // FontTools: take max k and compute v + mapping[k] - k
+                    return (value + self.to[idx - 1].into_inner()
+                        - self.from[idx - 1].into_inner())
+                    .into();
+                }
+
+                // This value is between two known values and we must lerp
+                let from_lhs = self.from[idx - 1].into_inner();
+                let from_rhs = self.from[idx].into_inner();
+                lerp(
+                    self.to[idx - 1].into_inner(),
+                    self.to[idx].into_inner(),
+                    (value - from_lhs) / (from_rhs - from_lhs),
+                )
+                .into()
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ordered_float::OrderedFloat;
+
+    use super::PiecewiseLinearMap;
+
+    #[test]
+    fn single_segment_map() {
+        // User likes to map wght 0..1000 onto 0..10
+        let from_to = vec![
+            (OrderedFloat(0_f32), OrderedFloat(0_f32)),
+            (OrderedFloat(10_f32), OrderedFloat(1000_f32)),
```

Comment:
IIUC you want to make wght=1000 (user scale) to 10 in internal design coordinates.. in that case you should swap the from and to in this example, because by definition `from` is user (fvar) coordintes and `to` is design coordinates.

Reply:
The other way around, 10 in internal design coordinates, e.g. what the source might have in something like a https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#location xvalue, to what fvar might have which IIUC we call user coordinates. I think perhaps the comment is written poorly, will update.